source/midi_stats.py

```diff
@@ -0,0 +1,500 @@
+
+
+class MidiStats:
+    def single_file_statistics(self, midi_file):
+        """
+        Compute statistics for a single midi path object.
+        """
+        # Some Midi files are corrupted and cannot be parsed by PrettyMIDI
+        try:
+            pm = PrettyMIDI(str(midi_file))
+        except Exception:
+            return None
+
+        # Compute statistics
+        statistics = {
+            # track md5 hash name without extension
+            "md5": track_name(midi_file),
+            # instruments
+            "n_instruments": n_instruments(pm),
+            "n_unique_instruments": n_unique_instruments(pm),
+            "instrument_names": instrument_names(pm),
+            "instrument_families": instrument_families(pm),
+            "number_of_instrument_families": number_of_instrument_families(pm),
+            # notes
+            "n_notes": number_of_notes(pm),
+            "n_unique_notes": number_of_unique_notes(pm),
+            "average_n_unique_notes_per_instrument": avg_number_of_unique_notes_per_instrument(
+                pm
+            ),
+            "average_note_duration": average_note_duration(pm),
+            "average_note_velocity": average_note_velocity(pm),
+            "average_note_pitch": average_note_pitch(pm),
+            "range_of_note_pitches": range_of_note_pitches(pm),
+            "average_range_of_note_pitches_per_instrument": average_range_of_note_pitches_per_instrument(
+                pm
+            ),
+            "number_of_note_pitch_classes": number_of_note_pitch_classes(pm),
+            "average_number_of_note_pitch_classes_per_instrument": average_number_of_note_pitch_classes_per_instrument(
+                pm
+            ),
+            "number_of_octaves": number_of_octaves(pm),
+            "average_number_of_octaves_per_instrument": average_number_of_octaves_per_instrument(
+                pm
+            ),
+            "number_of_notes_per_second": number_of_notes_per_second(pm),
+            "shortest_note_length": shortest_note_length(pm),
+            "longest_note_length": longest_note_length(pm),
+            # key signatures
+            "main_key_signature": main_key_signature(pm),  # hacky
+            "n_key_changes": n_key_changes(pm),
+            # tempo
+            "n_tempo_changes": n_tempo_changes(pm),
+            "tempo_estimate": average_tempo(pm),  # hacky
+            # time signatures
+            "main_time_signature": main_time_signature(pm),  # hacky
+            "all_time_signatures": all_time_signatures(pm),
+            "four_to_the_floor": four_to_the_floor(pm),
+            "n_time_signature_changes": n_time_signature_changes(pm),
+            # track length
+            "track_length_in_seconds": track_length_in_seconds(pm),
+            # lyrics
+            "lyrics_nb_words": lyrics_number_of_words(pm),
+            "lyrics_unique_words": lyrics_number_of_unique_words(pm),
+            "lyrics_bool": lyrics_boolean(pm),
+        }
```

Comment:
nice I like that function in dictionnary thingy. I always do more steps for clarity but here it is very nice and very clear.

Reply:
That's already how it is done in the Lakh tutorial! However they put the calculation functions directly in the lines. It's very compact but a bit messy